[PyTorch] Use plain old function pointer for RecordFunctionCallback

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#48629 [PyTorch] Use plain old function pointer for RecordFunctionCallback**
* #48620 [PyTorch] remove convenience RecordFunctionCallback interface

Nearly every non-test callsite doesn't need to capture any variables anyway, and this saves 48 bytes per callback.

Differential Revision: [D25135415](https://our.internmc.facebook.com/intern/diff/D25135415/)

swolchok committed:
[PyTorch] Use plain old function pointer for RecordFunctionCallback

Nearly every non-test callsite doesn't need to capture any variables anyway, and this saves 48 bytes per callback.

Differential Revision: [D25135415](https://our.internmc.facebook.com/intern/diff/D25135415/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Use plain old function pointer for RecordFunctionCallback"


Nearly every non-test callsite doesn't need to capture any variables anyway, and this saves 48 bytes per callback.

Differential Revision: [D25135415](https://our.internmc.facebook.com/intern/diff/D25135415/)

[ghstack-poisoned]

dhruvbird commented on aten/src/ATen/record_function.cpp:
```diff
@@ -252,10 +252,12 @@ class CallbackManager {
       }
       else {
-        rfcb.end()(rf, ctx.get());
+        if (rfcb.end()) {
+          rfcb.end()(rf, ctx.get());
+        }
```
I there a need to cache the return value from `rfcb.end()` to prevent a race condition where the value may change in between or will that not happen by construction?

kimishpatel commented:
It would be good to know benchmarking results of before and after.

dhruvbird commented:
Overall looks good to me. I assume that the loss of convenience (when testing) and the reliance on global/statics is something that is acceptable to everyone for the reduction in object size. I've left some comments on https://www.internalfb.com/diff/D25135415 for one of the fb-specific tests.

Is there a benchmark result for pre/post this change?

swolchok committed:
Update on "[PyTorch] Use plain old function pointer for RecordFunctionCallback"


Nearly every non-test callsite doesn't need to capture any variables anyway, and this saves 48 bytes per callback.

Differential Revision: [D25135415](https://our.internmc.facebook.com/intern/diff/D25135415/)

[ghstack-poisoned]

dhruvbird commented:
@kimishpatel  The benchmark results are in the test plan in the FB specific diff.

swolchok committed:
Update on "[PyTorch] Use plain old function pointer for RecordFunctionCallback"


Nearly every non-test callsite doesn't need to capture any variables anyway, and this saves 48 bytes per callback.

Differential Revision: [D25135415](https://our.internmc.facebook.com/intern/diff/D25135415/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Use plain old function pointer for RecordFunctionCallback"


Nearly every non-test callsite doesn't need to capture any variables anyway, and this saves 48 bytes per callback.

Differential Revision: [D25135415](https://our.internmc.facebook.com/intern/diff/D25135415/)

[ghstack-poisoned]

mruberry commented:
Unlanding. This appears to have broken numerous builds. Note that the Github CI also shows numerous failures. Example snippet:

```
Dec 15 04:34:01 ERROR 2020-12-15T04:27:50Z: sccache::server: Compilation failed: Output { status: ExitStatus(ExitStatus(256)), stdout: "", stderr: "\u{1b}[1m/var/lib/jenkins/workspace/test/cpp/jit/test_misc.cpp:773:3: \u{1b}[0m\u{1b}[0;1;31merror: \u{1b}[0m\u{1b}[1mno matching function for call to \'pushScopedCallback\'\u{1b}[0m\n  pushScopedCallback<at::RecordScope::FUNCTION, &fun_cnt>();\n\u{1b}[0;1;32m  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n\u{1b}[0m\u{1b}[1m/var/lib/jenkins/workspace/test/cpp/jit/test_misc.cpp:736:6: \u{1b}[0m\u{1b}[0;1;30mnote: \u{1b}[0mcandidate template ignored: invalid explicitly-specified argument for template parameter \'cnt\'\u{1b}[0m\nvoid pushScopedCallback() {\n\u{1b}[0;1;32m     ^\n\u{1b}[0m\u{1b}[1m/var/lib/jenkins/workspace/test/cpp/jit/test_misc.cpp:775:3: \u{1b}[0m\u{1b}[0;1;31merror: \u{1b}[0m\u{1b}[1mno matching function for call to \'pushScopedCallback\'\u{1b}[0m\n  pushScopedCallback<at::RecordScope::TORCHSCRIPT_FUNCTION, &ts_fun_cnt>();\n\u{1b}[0;1;32m  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n\u{1b}[0m\u{1b}[1m/var/lib/jenkins/workspace/test/cpp/jit/test_misc.cpp:736:6: \u{1b}[0m\u{1b}[0;1;30mnote: \u{1b}[0mcandidate template ignored: invalid explicitly-specified argument for template parameter \'cnt\'\u{1b}[0m\nvoid pushScopedCallback() {\n\u{1b}[0;1;32m     ^\n\u{1b}[0m\u{1b}[1m/var/lib/jenkins/workspace/test/cpp/jit/test_misc.cpp:777:3: \u{1b}[0m\u{1b}[0;1;31merror: \u{1b}[0m\u{1b}[1mno matching function for call to \'pushScopedCallback\'\u{1b}[0m\n  pushScopedCallback<at::RecordScope::USER_SCOPE, &user_scope_cnt>();\n\u{1b}[0;1;32m  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n\u{1b}[0m\u{1b}[1m/var/lib/jenkins/workspace/test/cpp/jit/test_misc.cpp:736:6: \u{1b}[0m\u{1b}[0;1;30mnote: \u{1b}[0mcandidate template ignored: invalid explicitly-specified argument for template parameter \'cnt\'\u{1b}[0m\nvoid pushScopedCallback() {\n\u{1b}[0;1;32m     ^\n\u{1b}[0m3 errors generated.\n" }

```

malfet commented:
@swolchok why this change was landed when it clearly introduced 15 new build failures?

swolchok commented:
> @swolchok why this change was landed when it clearly introduced 15 new build failures?

I've been trying to land it for a while, and I'm fairly sure I saw a clean CI run previously. Must have come in during rebasing. In particular, I would expect internal tests to have caught test_misc.cpp breaks and I'm confused that they didn't.

swolchok commented:
Looks like the test_misc changes in particular may have failed because of a difference between C++14 and C++17.
